test/Table_Tests/src/Database/Upload_Spec.enso

```diff
@@ -889,9 +889,169 @@ spec make_new_connection prefix persistent_connector=True =
 
+        Test.specify "will issue a friendly error if using in-memory table as target" <|
+            t1 = Table.new [["X", [1, 2, 3]]]
+            t2 = Table.new [["X", [1, 2, 3]]]
+            r1 = t1.update_rows t2
+            r1.should_fail_with Illegal_Argument
+            r1.to_display_text.should_contain "in-memory tables are immutable"
+
     Test.group prefix+"Appending a Database table to a Database table" <|
         test_table_append (database_table_builder "source-table") (database_table_builder "target-table")
 
+    test_table_delete source_table_builder target_table_builder =
+        Test.specify "should remove rows matching by key_columns" <|
+            table = target_table_builder [["student_id", [1, 2, 44, 100, 120]], ["first_name", ["Alice", "Bob", "Charlie", "David", "Eve"]], ["score", [100, 100, 44, 100, 120]]] primary_key=["student_id"]
+            key_values_to_delete = source_table_builder [["student_id", [44, 100]]]
+            # key columns should automatically be discovered by the primary key
+            affected_rows = table.delete_rows key_values_to_delete
+            affected_rows . should_equal 2
+            table.rows.map .to_vector . should_equal [[1, "Alice", 100], [2, "Bob", 100], [120, "Eve", 120]]
+
+        Test.specify "will require key_columns if no default can be used as no primary key is set" <|
+            table = target_table_builder [["X", [1, 2, 3]]] primary_key=[]
+            key_values_to_delete = source_table_builder [["X", [1, 2]]]
+
+            run_with_and_without_output <|
+                r1 = table.delete_rows key_values_to_delete
+                r1.should_fail_with Illegal_Argument
+                r1.to_display_text.should_contain "default value"
+
+            r2 = table.delete_rows key_values_to_delete key_columns=["X"]
+            Problems.assume_no_problems r2
+            r2.should_equal 2
+            table.at "X" . to_vector . should_equal [3]
+
+        Test.specify "does not fail if no rows matching the key_values_to_delete are found" <|
+            table = target_table_builder [["X", [1, 2, 3]]] primary_key=[]
+            key_values_to_delete = source_table_builder [["X", [4, 5]]]
+            r1 = table.delete_rows key_values_to_delete key_columns=["X"]
+            r1.should_equal 0
+            Problems.assume_no_problems r1
+            table.at "X" . to_vector . should_equal [1, 2, 3]
+
+            key_values_2 = source_table_builder [["X", [4, 3, 5]]]
+            r2 = table.delete_rows key_values_2 key_columns=["X"]
+            r2.should_equal 1
+            Problems.assume_no_problems r2
+            table.at "X" . to_vector . should_equal [1, 2]
+
+        Test.specify "should allow to use multiple columns as key" <|
+            table = target_table_builder [["X", [1, 2, 2, 3]], ["Y", ['a', 'b', 'c', 'd']]] primary_key=[]
+            keys = source_table_builder [["X", [2]], ["Y", ['b']]]
```

Comment:
Match multiple rows here too:

```suggestion
            table = target_table_builder [["X", [1, 2, 2, 3, 4, 4]], ["Y", ['a', 'b', 'c', 'd', 'e', 'f']]] primary_key=[]
            keys = source_table_builder [["X", [2, 4]], ["Y", ['b', 'f']]]
```

Reply:
What does this test change? I'm not sure if I understand what is the practical difference, but I'm happy to change it.

Reply:
It just exercises a more complicated case.

Reply:
ok